README.md
# hello-message

## Prerequisites
* SAM CLI - [Install the SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)
* [Python 3 installed](https://www.python.org/downloads/)
* Docker - [Install Docker community edition](https://hub.docker.com/search/?type=edition&offering=community)

```bash
pip install -r hello_world/requirements.txt --user
pip install -r tests/requirements.txt --user
```

### Start an MySQL container

Note: this MySQL stuff is a quick&dirty setup meant to spare time. Not to be used in real projects


```bash
docker run --name=acube \
  --restart on-failure \
  --env MYSQL_ALLOW_EMPTY_PASSWORD=true \
  -p 3307:3306 \
  -d mysql/mysql-server:5.7
```


### Setup the database
```bash
docker exec -it acube mysql -uroot
```

From the `mysql>` prompt, paste:
```
use mysql;
select host from user where user='root';
update user set host = '%' where user ='root';
flush privileges;
DROP DATABASE IF EXISTS acube;
CREATE DATABASE acube;
use acube;
CREATE TABLE IF NOT EXISTS stats(
  id INT AUTO_INCREMENT PRIMARY KEY,
  customer_id INT NOT NULL,
  type_id VARCHAR(255) NOT NULL,
  count INT,
  amount DECIMAL(10,3),
  UNIQUE KEY my_uniq_id (customer_id, type_id)
);
exit

```

Find the container's address:
```bash
docker inspect acube  | grep IPAddress
```

and put it in `hello_world/mysql_helper.py >> HOST variable`


## Build and local run
Build the application.

```bash
./scripts/build
```

Build and run in local with SAM. The application will respond on http://localhost:3000/

```bash
./scripts/run_local
```

## view logs
```bash
tail -f logfile.txt
```


## Tests
Tests are defined in the `tests` folder in this project.

```bash
./scripts/run_local  # if not running
./scripts/test       # on a different terminal
```





changelog.md
## last commit
- more readme
- complete one integration test

## ae29828847e175a7936c92ee740ca3c58b8d8a4d
- stored message into the database
- added quick and dirty mysql helper
- added pymysql to requirements
- implemented integration test

## d0c4e692b644135bd7dff069582667f8bf5f6532
- added build script
- renamed script unittest to test
- using docker for building
- written README.md
- gitignored logfile.txt

## 29bc330770bed94f7c2ca304c77ccd7c9ddb93aa
- added bash scripts: unit test, integration test(local), run locally

## 88fdd0840d227edcfc1c58de35a85aca0f988acc
- added changelog.md
- installed `requirements.txt` from both `hello_world/` and `test/`
- removed boto3, printed event's body
- cleaned unit and integration tests

## 1b9856b02b493c9450fae61009c9ebe272372925
- run `sam init` to download a basic template project

tests/unit/test_handler.py
import json

import pytest

from hello_world import app


@pytest.fixture()
def apigw_event():
    """ Generates API GW Event"""

    return {
        "body": """{
                    "customerId": 1,
                    "type": "A",
                    "amount": "0.012",
                    "uuid": "a596b362-08be-419f-8070-9c3055566e7c"
                }""",
        "resource": "/{proxy+}",
        "requestContext": {
            "resourceId": "123456",
            "apiId": "1234567890",
            "resourcePath": "/{proxy+}",
            "httpMethod": "POST",
            "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
            "accountId": "123456789012",
            "identity": {
                "apiKey": "",
                "userArn": "",
                "cognitoAuthenticationType": "",
                "caller": "",
                "userAgent": "Custom User Agent String",
                "user": "",
                "cognitoIdentityPoolId": "",
                "cognitoIdentityId": "",
                "cognitoAuthenticationProvider": "",
                "sourceIp": "127.0.0.1",
                "accountId": "",
            },
            "stage": "prod",
        },
        "queryStringParameters": {"foo": "bar"},
        "headers": {
            "Via": "1.1 08f323deadbeefa7af34d5feb414ce27.cloudfront.net (CloudFront)",
            "Accept-Language": "en-US,en;q=0.8",
            "CloudFront-Is-Desktop-Viewer": "true",
            "CloudFront-Is-SmartTV-Viewer": "false",
            "CloudFront-Is-Mobile-Viewer": "false",
            "X-Forwarded-For": "127.0.0.1, 127.0.0.2",
            "CloudFront-Viewer-Country": "US",
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
            "Upgrade-Insecure-Requests": "1",
            "X-Forwarded-Port": "443",
            "Host": "1234567890.execute-api.us-east-1.amazonaws.com",
            "X-Forwarded-Proto": "https",
            "X-Amz-Cf-Id": "aaaaaaaaaae3VYQb9jd-nvCd-de396Uhbp027Y2JvkCPNLmGJHqlaA==",
            "CloudFront-Is-Tablet-Viewer": "false",
            "Cache-Control": "max-age=0",
            "User-Agent": "Custom User Agent String",
            "CloudFront-Forwarded-Proto": "https",
            "Accept-Encoding": "gzip, deflate, sdch",
        },
        "pathParameters": {"proxy": "/examplepath"},
        "httpMethod": "POST",
        "stageVariables": {"baz": "qux"},
        "path": "/examplepath",
    }


def test_lambda_handler(apigw_event, mocker):

    ret = app.lambda_handler(apigw_event, "")
    data = json.loads(ret["body"])

    assert ret["statusCode"] == 200
    assert "message" in ret["body"]
    assert data["message"] == "event digested"


tests/integration/test_api_gateway.py
import json
from unittest import TestCase
import requests

"""
Test the Lambda function running on SAM
"""

class TestApiGateway(TestCase):
    # TODO put the endpoint in the environment
    api_endpoint: str = "http://127.0.0.1:3000/hello"

    def setUp(self) -> None:
        self.message = {
            "body": {"customerId": 2,"type": "B","amount": "0.012","uuid": "a596b362-08be-419f-8070-9c3055566e7c"},
            "resource": "/hello",
            "path": "/hello",
            "httpMethod": "GET",
            "isBase64Encoded": False,
            "queryStringParameters": {
                "foo": "bar"
            },
            "pathParameters": {
                "proxy": "/path/to/resource"
            },
            "stageVariables": {
                "baz": "qux"
            },
            "headers": {
                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
                "Accept-Encoding": "gzip, deflate, sdch",
                "Accept-Language": "en-US,en;q=0.8",
                "Cache-Control": "max-age=0",
                "CloudFront-Forwarded-Proto": "https",
                "CloudFront-Is-Desktop-Viewer": "true",
                "CloudFront-Is-Mobile-Viewer": "false",
                "CloudFront-Is-SmartTV-Viewer": "false",
                "CloudFront-Is-Tablet-Viewer": "false",
                "CloudFront-Viewer-Country": "US",
                "Host": "1234567890.execute-api.us-east-1.amazonaws.com",
                "Upgrade-Insecure-Requests": "1",
                "User-Agent": "Custom User Agent String",
                "Via": "1.1 08f323deadbeefa7af34d5feb414ce27.cloudfront.net (CloudFront)",
                "X-Amz-Cf-Id": "cDehVQoZnx43VYQb9j2-nvCh-9z396Uhbp027Y2JvkCPNLmGJHqlaA==",
                "X-Forwarded-For": "127.0.0.1, 127.0.0.2",
                "X-Forwarded-Port": "443",
                "X-Forwarded-Proto": "https"
            },
            "requestContext": {
                "accountId": "123456789012",
                "resourceId": "123456",
                "stage": "prod",
                "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
                "requestTime": "09/Apr/2015:12:34:56 +0000",
                "requestTimeEpoch": 1428582896000,
                "identity": {
                "cognitoIdentityPoolId": None,
                "accountId": None,
                "cognitoIdentityId": None,
                "caller": None,
                "accessKey": None,
                "sourceIp": "127.0.0.1",
                "cognitoAuthenticationType": None,
                "cognitoAuthenticationProvider": None,
                "userArn": None,
                "userAgent": "Custom User Agent String",
                "user": None
                },
                "path": "/prod/hello",
                "resourcePath": "/hello",
                "httpMethod": "POST",
                "apiId": "1234567890",
                "protocol": "HTTP/1.1"
            }
}


    def test_api_gateway(self):
        """
        Call the API Gateway endpoint with an event and check the response
        """
        response = requests.get(self.api_endpoint, json=self.message)
        self.assertDictEqual(response.json(), {"message": "event digested"})


python/message_digest/index.py
import json
import pymysql.cursors
import mysql_helper

def handler(event, context):
    """Message digesting Lambda function

    Parameters
    ----------
    event: dict, required
        API Gateway Lambda Proxy Input Format

        Event doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format

    context: object, required
        Lambda Context runtime methods and attributes

        Context doc: https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html

    Returns
    ------
    API Gateway Lambda Proxy Output Format: dict

        Return doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html
    """
    
    try:
        body = event['body']['body']
        customerId = body['customerId']
        messageType = body['type']
        amount = body['amount']
    except:
        return {
            "statusCode": 400,
            "body": {"message": "invalid event"},
    }

    try:
        mysql_helper.update(customerId, messageType, amount)
    except:
        return {
            "statusCode": 400,
            "body": {"message": "database error"},
    }

    return {
        "statusCode": 200,
        "body": {"message": "event digested"},
    }


python/message_digest/mysql_helper.py
import pymysql.cursors
import os

# TODO as this is a quick and dirty mysql call, params are in the code. They must come from the environment
USER = os.environ['MYSQL_USER']
DB = os.environ['MYSQL_DATABASE']

def get_connection():
  connection = pymysql.connect(
    user=USER,
    database=DB,
    charset='utf8mb4',
    cursorclass=pymysql.cursors.DictCursor)
  return connection

def update(customerId, messageType, amount):
  """
  Inserts a new row in the database if customer_id + type_id does not exist
  Otherwise increments the count field by one and add the amount to the existing amout
  """
  connection = get_connection()
  with connection:
      with connection.cursor() as cursor:
          # Create a new record
          sql = """
          INSERT INTO stats 
            (customer_id, type_id, count, amount) 
          VALUES
            (%(customerId)s, %(messageType)s, 1, %(amount)s)
          ON DUPLICATE KEY UPDATE
            count = count + 1,
            amount = amount + %(amount)s
          """
          cursor.execute(sql, {"customerId": customerId, "messageType": messageType, "amount": amount})

      connection.commit()

if __name__ == "__main__":
  print(update(1, 'a', '1.23'))

python/message_digest/secret_helper.py
import boto3
import base64
from botocore.exceptions import ClientError


def get_secret():

    secret_name = "digest-secret"
    region_name = "us-east-1"

    # Create a Secrets Manager client
    session = boto3.session.Session()
    client = session.client(
        service_name='secretsmanager',
        region_name=region_name
    )

    try:
        get_secret_value_response = client.get_secret_value(
            SecretId=secret_name
        )
    except ClientError as e:
        print( e.response['Error']['Code'] )
        raise e
    else:
        # Decrypts secret using the associated KMS key.
        # Depending on whether the secret is a string or binary, one of these fields will be populated.
        if 'SecretString' in get_secret_value_response:
            secret = get_secret_value_response['SecretString']
        else:
            decoded_binary_secret = base64.b64decode(get_secret_value_response['SecretBinary'])
            


hello_world/app.py
import json
import mysql_helper

def lambda_handler(event, context):
    """Message digesting Lambda function

    Parameters
    ----------
    event: dict, required
        API Gateway Lambda Proxy Input Format

        Event doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format

    context: object, required
        Lambda Context runtime methods and attributes

        Context doc: https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html

    Returns
    ------
    API Gateway Lambda Proxy Output Format: dict

        Return doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html
    """

    try:
        body = json.loads(event['body'])['body']
        customerId = body['customerId']
        messageType = body['type']
        amount = body['amount']
    except:
        return {
            "statusCode": 400,
            "body": json.dumps({
                "message": "invalid event"
        }),
    }

    try:
        mysql_helper.update(customerId, messageType, amount)
    except:
        return {
            "statusCode": 400,
            "body": json.dumps({
                "message": "database error"
        }),
    }

    return {
        "statusCode": 200,
        "body": json.dumps({
            "message": "event digested"
        }),
    }
